golang/cosmos/x/swingset/genesis.go

```diff
@@ -131,7 +182,17 @@ func (eventHandler swingStoreGenesisEventHandler) OnExportRetrieved(provider kee
 
 	artifactsProvider := keeper.SwingStoreExportProvider{
 		GetExportDataReader: func() (agoric.KVEntryReader, error) {
-			return nil, nil
+			exportDataIterator := eventHandler.swingStore.Iterator(nil, nil)
+			kvReader := agoric.NewKVIteratorReader(exportDataIterator)
+			eventHandler.hasher.Reset()
+			encoder := json.NewEncoder(eventHandler.hasher)
+			encoder.SetEscapeHTML(false)
+
+			return agoric.NewKVHookingReader(kvReader, func(entry agoric.KVEntry) error {
+				return encoder.Encode(entry)
```

Comment:
Hm, I can see that the "what do we hash" question is tied into the "what does golang encode" question. Giving golang the responsibility for framing/unframing, and then JSON-encoding the key/value pairs for hashing, is safe against concatenation/framing-confusion attacks, even if it's vulnerable to some hypothetical non-injectiveness in JSON.

If we were to, e.g., skip JSON, and just feed the key into the hasher, then feed the value into the hasher, then repeat for each pair: that would be vulnerable to framing confusion (`{ ab: 'cd', ef: 'gh'}` would hash the same as `{a: 'b', c: 'de', fg: 'h'}`). Without some kind of encoding/escaping, there's no way to avoid that (simply using a newline as a framing separator could still be confused with newlines in the keys or values).

One option would be to JSON-encode the key/value pairs first, then have the golang codec (struct? proto?) encode the resulting strings (one string per entry, not one for the key and a second for the value). Then the hasher could be fed those `JSON.stringify([key,value])` strings, which would be safely escaped, and unambiguously framed. It would look weird, eschewing the built-in encoding support in favor of manual JSON, but it would meet my maxim of "hash (sign) what you say, only use what you hash (sign)". Down with "canonical encoding" in secure protocols!

Reply:
> Then the hasher could be fed those `JSON.stringify([key,value])` strings, which would be safely escaped, and unambiguously framed.

That's exactly what this does. Bonus is that the hash ends up strictly equal to hashing the resulting file, which is a concatenation of all the JSON stringified entries terminated by new lines

Reply:
Just to make sure I was clear, you wouldn't extract `key` and `value` from the golang-encoded data and then feed `JSON.stringify([key,value])` to the hasher: that's the vulnerable-to-JSON-weirdness order. Rather, you'd feed `JSON.stringify([key,value])` to the golang *encoder*, and have golang do whatever additional/native encoding it wants on that single-string-per-entry. Then, on the decoding side, you'd first hash that single-string-per-entry, then you do `JSON.parse` to get the key/value that you need to act upon. The rule could also be summarized as "no encoding during decode".

But I hear you, we're fighting against the golang codec pattern, which has already frustrated me in cosmos txn generation/signing. And we're dependent upon their codec being possible to use securely, which isn't a given (the confusion between `nil` and an empty list comes to mind).

Reply:
The missing bit is that the incoming stream of key value pairs may not be JSON encoded in the first place. It happens that's how we currently feed golang when dealing with export/import, but that may not always be the case.